server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java

```diff
@@ -0,0 +1,359 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.pipeline.movfn;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+/**
+ * Provides a collection of static utility methods that can be referenced from MovingFunction script contexts
+ */
+public class MovingFunctions {
+
+    /**
+     * Find the maximum value in a window of values.
+     * If all values are missing/null/NaN, the return value will be NaN
+     */
+    public static double max(Collection<Double> values) {
```

Comment:
I think `double[]` should be doable.  :+1: 

I didn't do that originally to avoid unboxing the internal `EvictingQueue` into an array since it wouldn't be necessary most of the time.  And I avoided specifying a `Queue` as the type since the user might want to use these functions outside of the values we provide (or in addition to).

But I didn't think about the unordered nature of `Collection`, that's definitely an issue.  Lemme poke at it tomorrow with `double[]`, I think that'll work fine.

_Sidebar_ 
Alternatively, I could replace the evicting queue with a primitive-based ring buffer.  I don't think that's any better though.  We'd still have to arrayCopy the relevant slice into the script since there are null/NaNs to deal with, and we probably shouldn't make the method signature some weird ring buffer class :)

> Note that doc['fieldname'] returns an object that implements List, but we could make this easier by adding an asArray() method to DoubleScriptDocValues?

Ah interesting.  I'm not sure it's needed... do we think users will plop a field's value into the functions instead of the values?  I was thinking users might tweak the values before a function, or write their own based on the values... but not use an entire field in place of the provided values.  Thoughts?

Reply:
Forget what I said about ScriptDocValues. I was thinking this was like scripted metric agg, which is a SearchScript, but you did this correctly (no getDoc()). So no need to do anything with the doc values accessors.